### PR TITLE
qt: fix shutdown on MacOS

### LINKED ISSUE #2362: Segfault when quitting Mac app while loading blocks
# Bug Report

**Current behavior**

Start the gridcoin client on Mac and quit while the "Blocks Loaded" status percentage is small (5% loaded crashed, 11% loaded did not). I'm able to reproduce this. 

**Expected behavior**

Client quits normally without a segfault

**Steps to reproduce:**

Start gridcoin client, right click on the Gridcoin icon in the dock, select "Quit", while in the early stage of loading blocks (less than 5% loaded). NB that I can't repro this using Cmd-Q, only by selecting Quit from the dock icon. I then get a crash report from MacOS with error logs that say a segfault happened. 

**Screenshots**

**Gridcoin version**

version 5.3.2-unk

**Machine specs**
- OS: macOS
- CPU: x86_64
- RAM: 8 GB
- Disk size: 256 GB
- Disk Type (HD/SSD): SSD

See pastebin debug info for more machine info. I'm on MacOS Monterey beta 10, fwiw. I'll try and reproduce with Monterey 12.0.1 which was just released today. 

**Extra information**

Error logs reported by macOS:
https://pastebin.com/zbFxKvnT

### PR DESCRIPTION
**Pull request by div72**
Fixes #2362.

### CODE REVIEW

**Comment by div72** [comment]
Unintended side effect: hooking applicationShouldTerminate also delays the shutdown/reboot until the wallet is closed.

File: src/qt/macdockiconhandler.mm [hunk 1]
```diff
@@ -44,6 +46,25 @@ - (void)handleDockClickEvent:(NSAppleEventDescriptor*)event withReplyEvent:(NSAp
...
+    Class delClass = (Class)[[[NSApplication sharedApplication] delegate] class];
+    SEL closeHandle = sel_registerName("applicationShouldTerminate:");
+    if (class_getInstanceMethod(delClass, closeHandle)) {
+        class_replaceMethod(delClass, closeHandle, (IMP)MacDockShutdownHandler::handleShutdown, "B@:");
+    } else {
+        class_addMethod(delClass, closeHandle, (IMP)MacDockShutdownHandler::handleShutdown,"B@:");
+    }
+}
+
+MacDockShutdownHandler::~MacDockShutdownHandler() {}
```
Comment (denravonska) [inline src/qt/macdockiconhandler.mm]: This one can be nuked, unless it's ported from Bitcoin. If so, keep it to stay in line with upstream :)

**Comment by ghost** [comment]
Tested OK on latest MacOS on M1.

File: src/qt/macdockiconhandler.mm (hunk 1 above)
Comment (div72) [reply src/qt/macdockiconhandler.mm]: I was planning on de-hooking but the application is already exiting, so why bother? Nuked.

**Review by jamescowens — APPROVED** [review]
ACK

**Pull request merged**